go/store/prolly/tree/json_chunker.go

```diff
@@ -0,0 +1,260 @@
+	for {
+		j.appendJsonToBuffer(jsonBytes)
+		j.processBuffer(ctx)
+		if j.jScanner.jsonBuffer == nil {
+			// Advance the cursor so that we don't re-insert the current key when finalizing the chunker.
+			j.jCur.cur.advance(ctx)
+			return j.chunker.Done(ctx)
+		}
+		err := cur.advance(ctx)
+		if err != nil {
+			return Node{}, err
+		}
+		if !cur.Valid() {
+			// We reached the end of the tree.
+			err := j.createNewLeafChunk(ctx, endOfDocumentKey, j.jScanner.jsonBuffer)
+			if err != nil {
+				return Node{}, err
+			}
+			return j.chunker.Done(ctx)
+		}
+		jsonBytes = cur.currentValue()
+	}
+}
+
+// createNewLeafChunk writes a new Blob to the nodestore, and updates the parent chunker.
+// Do not call this method directly. It should only get called from within this file.
+func (j *JsonChunker) createNewLeafChunk(ctx context.Context, key, value []byte) error {
+	blobSerializer := message.NewBlobSerializer(j.ns.Pool())
+	msg := blobSerializer.Serialize(nil, [][]byte{value}, []uint64{1}, 0)
+	node, err := NodeFromBytes(msg)
+	if err != nil {
+		return err
+	}
+	addr, err := j.ns.Write(ctx, node)
+	if err != nil {
+		return err
+	}
+	// Copy the key when adding it to the chunker.
+	return j.chunker.AddPair(ctx, bytes.Clone(key), addr[:])
+}
+
+// appendJsonWithoutSplitting writes JSON to the buffer, but prevents the new JSON from later being scanned for chunk boundaries.
+// This is useful in situations where we know that crossing a boundary is not possible and know what the resulting scanner state will be.
+// Do not call this method directly. It should only get called from within this file.
+func (j *JsonChunker) appendJsonWithoutSplitting(jsonBytes []byte) {
+	j.appendJsonToBuffer(jsonBytes)
+	j.jScanner.skipBytes(len(jsonBytes))
+}
+
+// appendJsonToBuffer writes JSON to the buffer, which will be checked for chunk boundaries the next time processBuffer is called.
+// Do not call this method directly. It should only get called from within this file.
+func (j *JsonChunker) appendJsonToBuffer(jsonBytes []byte) {
+	if j.jScanner.jsonBuffer == nil {
+		j.jScanner.jsonBuffer = jsonBytes
+	} else {
+		j.jScanner.jsonBuffer = append(j.jScanner.jsonBuffer, jsonBytes...)
+	}
+}
+
+// processBuffer reads all new additions added by appendJsonToBuffer, and determines any new chunk boundaries.
+// Do not call this method directly. It should only get called from within this file.
+func (j *JsonChunker) processBuffer(ctx context.Context) error {
+	for j.jScanner.AdvanceToNextLocation() != io.EOF {
+		key := j.jScanner.currentPath.key
+		value := j.jScanner.jsonBuffer[:j.jScanner.valueOffset]
+		if crossesBoundary(key, value) {
+			err := j.createNewLeafChunk(ctx, key, value)
+			if err != nil {
+				return err
+			}
+			var newBuffer []byte
+			if !j.jScanner.atEndOfChunk() {
+				newBuffer = j.jScanner.jsonBuffer[j.jScanner.valueOffset:]
+			}
+			newScanner := ScanJsonFromMiddle(newBuffer, j.jScanner.currentPath)
+			j.jScanner = &newScanner
+		}
+	}
```

Comment:
Is it more clear to just keep the start offset explicitly rather than resetting j.jScanner each time?

```go
	startOffset := 0
	for j.jScanner.AdvanceToNextLocation() != io.EOF {
		key := j.jScanner.currentPath.key
		value := j.jScanner.jsonBuffer[startOffset:j.jScanner.valueOffset]
		if crossesBoundary(key, value) {
			err := j.createNewLeafChunk(ctx, key, value)
			if err != nil {
				return err
			}
			startOffset = j.jScanner.valueOffset
		}
	}
	j.jScanner = ScanJsonFromMiddle(j.jScanner.jsonBuffer[startOffset:], j.jScanner.currentPath)
	return nil
```